Could not find path in MethodInvocationTree

```
$ cat <<EOF > C.java
import java.util.Arrays;
class C {
    void f() {
        Arrays.asList(0, 0, "", Arrays.asList(new Object[0]));
    }
}
EOF
$ java -jar $CHECKERFRAMEWORK/checker/dist/checker.jar\
 -Xbootclasspath/p:$CHECKERFRAMEWORK/checker/dist/jdk8.jar\
 -XDignore.symbol.file=true -Xmaxerrs 20000 -Xmaxwarns 20000\
 -source 8 -target 8 -encoding ascii\
 -cp ../../build:$JSR308/annotation-tools/annotation-file-utilities.jar\
 -processor nullness -Anocheckjdk -Aignorejdkastub\
 -AuseDefaultsForUncheckedCode=source -Awarns -AprintErrorStack C.java
error: SourceChecker.typeProcess: unexpected Throwable (AssertionError) while processing C.java; message: Could not find path in MethodInvocationTree.
  treePath=com.sun.source.util.TreePath@4f704591
  methodInvocation=Arrays.asList(0, 0, "", Arrays.asList(new Object[0]))
  Compilation unit: C.java
  Exception: java.lang.AssertionError: Could not find path in MethodInvocationTree.
  treePath=com.sun.source.util.TreePath@4f704591
  methodInvocation=Arrays.asList(0, 0, "", Arrays.asList(new Object[0])); Stack trace: org.checkerframework.framework.util.typeinference.TypeArgInferenceUtil.assignedTo(TypeArgInferenceUtil.java:158)
  org.checkerframework.framework.util.typeinference.DefaultTypeArgumentInference.getAssignedTo(DefaultTypeArgumentInference.java:125)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory$KeyForTypeArgumentInference.getAssignedTo(KeyForAnnotatedTypeFactory.java:896)
  org.checkerframework.framework.util.typeinference.DefaultTypeArgumentInference.inferTypeArgs(DefaultTypeArgumentInference.java:99)
  org.checkerframework.framework.util.AnnotatedTypes.findTypeArguments(AnnotatedTypes.java:693)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1817)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1803)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1131)
  org.checkerframework.checker.nullness.KeyForAnnotatedTypeFactory.methodFromUse(KeyForAnnotatedTypeFactory.java:184)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:221)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:52)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1171)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:952)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:214)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:887)
  org.checkerframework.checker.nullness.KeyForTransfer.visitMethodInvocation(KeyForTransfer.java:58)
  org.checkerframework.checker.nullness.KeyForTransfer.visitMethodInvocation(KeyForTransfer.java:30)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:81)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:362)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:220)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:893)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:861)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:805)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1090)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.addComputedTypeAnnotations(GenericAnnotatedTypeFactory.java:1057)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.addComputedTypeAnnotations(GenericAnnotatedTypeFactory.java:1037)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:959)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:2118)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:300)
  org.checkerframework.common.basetype.BaseTypeVisitor.visitClass(BaseTypeVisitor.java:174)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:70)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:942)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:448)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:442)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:209)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
error: SourceChecker.typeProcess: unexpected Throwable (AssertionError) while processing C.java; message: Could not find path in MethodInvocationTree.
  treePath=com.sun.source.util.TreePath@3af0a9da
  methodInvocation=Arrays.asList(0, 0, "", Arrays.asList(new Object[0]))
  Compilation unit: C.java
  Exception: java.lang.AssertionError: Could not find path in MethodInvocationTree.
  treePath=com.sun.source.util.TreePath@3af0a9da
  methodInvocation=Arrays.asList(0, 0, "", Arrays.asList(new Object[0])); Stack trace: org.checkerframework.framework.util.typeinference.TypeArgInferenceUtil.assignedTo(TypeArgInferenceUtil.java:158)
  org.checkerframework.framework.util.typeinference.DefaultTypeArgumentInference.getAssignedTo(DefaultTypeArgumentInference.java:125)
  org.checkerframework.framework.util.typeinference.DefaultTypeArgumentInference.inferTypeArgs(DefaultTypeArgumentInference.java:99)
  org.checkerframework.framework.util.AnnotatedTypes.findTypeArguments(AnnotatedTypes.java:693)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1817)
  org.checkerframework.framework.type.AnnotatedTypeFactory.methodFromUse(AnnotatedTypeFactory.java:1803)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.methodFromUse(GenericAnnotatedTypeFactory.java:1131)
  org.checkerframework.checker.nullness.NullnessAnnotatedTypeFactory.methodFromUse(NullnessAnnotatedTypeFactory.java:283)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:221)
  org.checkerframework.framework.type.TypeFromExpressionVisitor.visitMethodInvocation(TypeFromExpressionVisitor.java:52)
  com.sun.tools.javac.tree.JCTree$JCMethodInvocation.accept(JCTree.java:1477)
  com.sun.source.util.SimpleTreeVisitor.visit(SimpleTreeVisitor.java:53)
  org.checkerframework.framework.type.TypeFromTree.fromExpression(TypeFromTree.java:36)
  org.checkerframework.framework.type.AnnotatedTypeFactory.fromExpression(AnnotatedTypeFactory.java:1171)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:952)
  org.checkerframework.framework.flow.CFAbstractTransfer.getValueFromFactory(CFAbstractTransfer.java:214)
  org.checkerframework.framework.flow.CFAbstractTransfer.visitMethodInvocation(CFAbstractTransfer.java:887)
  org.checkerframework.checker.initialization.InitializationTransfer.visitMethodInvocation(InitializationTransfer.java:211)
  org.checkerframework.checker.nullness.NullnessTransfer.visitMethodInvocation(NullnessTransfer.java:221)
  org.checkerframework.checker.nullness.NullnessTransfer.visitMethodInvocation(NullnessTransfer.java:61)
  org.checkerframework.dataflow.cfg.node.MethodInvocationNode.accept(MethodInvocationNode.java:81)
  org.checkerframework.dataflow.analysis.Analysis.callTransferFunction(Analysis.java:362)
  org.checkerframework.dataflow.analysis.Analysis.performAnalysis(Analysis.java:220)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:893)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.analyze(GenericAnnotatedTypeFactory.java:861)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.performFlowAnalysis(GenericAnnotatedTypeFactory.java:805)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.checkAndPerformFlowAnalysis(GenericAnnotatedTypeFactory.java:1090)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.addComputedTypeAnnotations(GenericAnnotatedTypeFactory.java:1057)
  org.checkerframework.checker.nullness.NullnessAnnotatedTypeFactory.addComputedTypeAnnotations(NullnessAnnotatedTypeFactory.java:200)
  org.checkerframework.framework.type.GenericAnnotatedTypeFactory.addComputedTypeAnnotations(GenericAnnotatedTypeFactory.java:1037)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:959)
  org.checkerframework.framework.type.AnnotatedTypeFactory.getAnnotatedType(AnnotatedTypeFactory.java:2118)
  org.checkerframework.checker.initialization.InitializationVisitor.visitClass(InitializationVisitor.java:271)
  org.checkerframework.checker.initialization.InitializationVisitor.visitClass(InitializationVisitor.java:62)
  com.sun.tools.javac.tree.JCTree$JCClassDecl.accept(JCTree.java:720)
  com.sun.source.util.TreePathScanner.scan(TreePathScanner.java:50)
  org.checkerframework.framework.source.SourceVisitor.visit(SourceVisitor.java:70)
  org.checkerframework.framework.source.SourceChecker.typeProcess(SourceChecker.java:942)
  org.checkerframework.common.basetype.BaseTypeChecker.typeProcess(BaseTypeChecker.java:448)
  org.checkerframework.javacutil.AbstractTypeProcessor$AttributionTaskListener.finished(AbstractTypeProcessor.java:209)
  com.sun.tools.javac.api.ClientCodeWrapper$WrappedTaskListener.finished(ClientCodeWrapper.java:681)
  com.sun.tools.javac.api.MultiTaskListener.finished(MultiTaskListener.java:111)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1342)
  com.sun.tools.javac.main.JavaCompiler.flow(JavaCompiler.java:1296)
  com.sun.tools.javac.main.JavaCompiler.compile2(JavaCompiler.java:901)
  com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:860)
  com.sun.tools.javac.main.Main.compile(Main.java:523)
  com.sun.tools.javac.main.Main.compile(Main.java:381)
  com.sun.tools.javac.main.Main.compile(Main.java:370)
  com.sun.tools.javac.main.Main.compile(Main.java:361)
  com.sun.tools.javac.Main.compile(Main.java:56)
  com.sun.tools.javac.Main.main(Main.java:42)
2 errors
```

--- a/framework/src/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
+++ b/framework/src/org/checkerframework/framework/util/typeinference/TypeArgInferenceUtil.java
@@ -145,35 +145,7 @@ public class TypeArgInferenceUtil {
                 return null;
             ExecutableElement methodElt = TreeUtils.elementFromUse(methodInvocation);
             AnnotatedTypeMirror receiver = atypeFactory.getReceiverType(methodInvocation);
-            AnnotatedExecutableType method = AnnotatedTypes.asMemberOf(types, atypeFactory, receiver, methodElt);
-            int treeIndex = -1;
-            for (int i = 0; i < method.getParameterTypes().size(); ++i) {
-                ExpressionTree argumentTree = methodInvocation.getArguments().get(i);
-                if (isArgument(path, argumentTree)) {
-                    treeIndex = i;
-                    break;
-                }
-            }
-            assert treeIndex != -1 :  "Could not find path in MethodInvocationTree.\n"
-                                      + "treePath=" + path.toString() + "\n"
-                                      + "methodInvocation=" + methodInvocation;
-
-            if (treeIndex == -1) {
-                return null;
-            }
-
-            final AnnotatedTypeMirror paramType = method.getParameterTypes().get(treeIndex);
-
-            // Examples like this:
-            // <T> T outMethod()
-            // <U> void inMethod(U u);
-            // inMethod(outMethod())
-            // would require solving the constraints for both type argument inferences simultaneously
-            if (paramType == null || containsUninferredTypeParameter(paramType, method)) {
-                return null;
-            }
-
-            return paramType;
+            return assignedToExecutable(atypeFactory, path, methodElt, receiver, methodInvocation.getArguments());
         } else if (assignmentContext instanceof NewArrayTree) {
             //TODO: I left the previous implementation below, it definitely caused infinite loops if you
             //TODO: called it from places like the TreeAnnotator
@@ -190,24 +162,8 @@ public class TypeArgInferenceUtil {
             NewClassTree newClassTree = (NewClassTree) assignmentContext;
             ExecutableElement constructorElt = InternalUtils.constructor(newClassTree);
             AnnotatedTypeMirror receiver = atypeFactory.fromNewClass(newClassTree);
-            AnnotatedExecutableType constructor = AnnotatedTypes.asMemberOf(types, atypeFactory, receiver, constructorElt);
-            int treeIndex = -1;
-            for (int i = 0; i < constructor.getParameterTypes().size(); ++i) {
-                ExpressionTree argumentTree = newClassTree.getArguments().get(i);
-                if (isArgument(path, argumentTree)) {
-                    treeIndex = i;
-                    break;
-                }
-            }
-
-            assert treeIndex != -1 :  "Could not find path in NewClassTree.\n"
-                    + "treePath=" + path.toString() + "\n"
-                    + "methodInvocation=" + newClassTree;
-            if (treeIndex == -1) {
-                return null;
-            }
-
-            return constructor.getParameterTypes().get(treeIndex);
+            return assignedToExecutable(atypeFactory, path, constructorElt, receiver,
+                    newClassTree.getArguments());
         } else if (assignmentContext instanceof ReturnTree) {
             HashSet<Kind> kinds = new HashSet<>(Arrays.asList(Kind.LAMBDA_EXPRESSION, Kind.METHOD));
             Tree enclosing = TreeUtils.enclosingOfKind(path, kinds);
@@ -227,6 +183,41 @@ public class TypeArgInferenceUtil {
 
         ErrorReporter.errorAbort("AnnotatedTypes.assignedTo: shouldn't be here!");
         return null; // dead code
+    }
+
+    private static AnnotatedTypeMirror assignedToExecutable(AnnotatedTypeFactory atypeFactory,
+                                                            TreePath path, ExecutableElement methodElt,
+                                                            AnnotatedTypeMirror receiver,
+                                                            List<? extends ExpressionTree> arguments) {
+        AnnotatedExecutableType method =
+                AnnotatedTypes.asMemberOf(atypeFactory.getContext().getTypeUtils(), atypeFactory, receiver, methodElt);
+        int treeIndex = -1;
+        for (int i = 0; i < arguments.size(); ++i) {
+            ExpressionTree argumentTree = arguments.get(i);
+            if (isArgument(path, argumentTree)) {
+                treeIndex = i;
+                break;
+            }
+        }
+        assert treeIndex != -1 :  "Could not find path in MethodInvocationTree.\n"
+                                  + "treePath=" + path.toString();
+        final AnnotatedTypeMirror paramType;
+        if (treeIndex >= method.getParameterTypes().size() && methodElt.isVarArgs()) {
+            paramType = method.getParameterTypes().get( method.getParameterTypes().size() -1);
+        } else {
+            paramType = method.getParameterTypes().get(treeIndex);
+        }
+
+        // Examples like this:
+        // <T> T outMethod()
+        // <U> void inMethod(U u);
+        // inMethod(outMethod())
+        // would require solving the constraints for both type argument inferences simultaneously
+        if (paramType == null || containsUninferredTypeParameter(paramType, method)) {
+            return null;
+        }
+
+        return paramType;
     }
 
     /**

--- a/framework/tests/all-systems/Issue808.java
+++ b/framework/tests/all-systems/Issue808.java
@@ -1,0 +1,17 @@
+// Test case for Issue 808
+// https://github.com/typetools/checker-framework/issues/808
+import java.util.Arrays;
+class Issue808 {
+    void f() {
+       Arrays.asList(0, 0, "", Arrays.asList(new Object[0]));
+       foo(new Object(), bar());
+       new Issue808(bar());
+    }
+
+    <T> T bar() {
+        throw new RuntimeException();
+    }
+
+    void foo(Object... param) {}
+    Issue808(Object... param) {}
+}

